### DISCUSSION TITLE
How do I create a zod schema that's of File type?

### DISCUSSION THREAD

**Discussion by ljunio** [colinhacks/zod#801] [Q&A]
I'm trying to create a zod object that contains a File type. What would be the best schema type for this?

**Comment by scotttrinh** [comment]
`z.instanceof(File)` should work.

[From the README:](https://github.com/colinhacks/zod#instanceof)

> You can use `z.instanceof` to check that the input is an instance of a class. This is useful to validate inputs against classes that are exported from third-party libraries.

**Reply by ljunio** [reply]
Thanks for the response @scotttrinh. I ended up using `z.instanceof(File)`, though, after upload, I'm getting the error "Input not instance of File" even though I'm uploading a 2.8kb png. Am I missing something?
```ts
const schema = z.object({
  file: z.instanceof(File);
});

**Reply by scotttrinh** [reply]
oh I think I saw another issue a while back about this: the schema parser does actually return an object rather than passing through the input type unchanged. Let me see if I can find the fix for that.

**Reply by scotttrinh** [reply]
Ahh, it doesn't look like we got a resolution there. Let me investigate if there is anything we can do there.

**Reply by scotttrinh** [reply]
Nothing is immediately obvious while looking at the parsing code. It _should_ be passing the data without touching it. Before digging in further, can you verify that you're on the latest Zod version, or just share what version you are on if not the latest?

**Reply by ljunio** [reply]
@scotttrinh just to confirm, does that mean that the schema parser (underneath `.instanceof`) is actually changing the type to an object instead of leaving it be?

And yeah, I can confirm that I'm on the latest zod version.

**Reply by scotttrinh** [reply]
I'm not sure if it's the `instanceof` parser or the `object` parser that is wrapping it, but _something_ is seemingly changing the prototype to be `Object` instead of `File` in the parsed data. A workaround is to use the input data rather than the parsed data, but I consider this a hack around a bug rather than a suggested approach.

I know one tricky bit is that the `File` class in browser contexts can be easy to lose when passed around to different closures. I seem to recall having some issues with that some years ago when writing an Electron app that dealt with file uploads. Might be a tricky one to get right, tbh. Going to work on a minimal repro so we can test some assumptions.
